src/Microsoft.Net.Http.Headers/HeaderUtilities.cs

```diff
@@ -480,6 +479,100 @@ public static unsafe bool TryParseInt64(StringSegment value, out long result)
             }
         }
 
+        // Fast RFC2616 3.9 Quality value parser (and without memory allocation)
+        internal static bool TryParseQualityDouble(string input, int startIndex, out double quality, out int length)
+        {
+            quality = 0;
+            length = 0;
+
+            var inputLength = input.Length;
+            var current = startIndex;
+            var limit = startIndex + 9;         // Accept a maximum of '9' chars to avoid 'int' overflows (RFC2616 3.9 even accepts only 5 chars)
+            var stop = false;
+
+            var intPart = 0;
+            var decPart = 0;
+            var decPow = 1;
+
+            if (current >= inputLength)
+            {
+                return false;
+            }
+
+            // We are more restrictive for the first char (we accept only digits)
+            var ch = input[current];
+
+            if (ch >= '0' && ch <= '9')
+            {
+                intPart = ch - '0';
+                current++;
+            }
+            else
+            {
+                // The RFC doesn't allow decimal values starting with dot. I.e. value ".123" is invalid. It must be in the
+                // form "0.123".
+                return false;
+            }
+
+            // Other chars
+            while (current < inputLength && !stop)
+            {
+                ch = input[current];
+
+                if (current >= limit)
+                {
+                    return false;
+                }
+                else if (ch >= '0' && ch <= '9')
+                {
+                    intPart = intPart * 10 + ch - '0';
```

Comment:
Similar to the previous comment, these are not valid doubles so parsing should fail.